### PR TITLE
Add Python interface for compilation and one single interface for compilation and running process

### PR DESCRIPTION
**Pull request by Jiaqing-ASU**
An ONNX model can be compiled directly using the `onnx-mlir -O3 --EmitLib` command. The resulting library can then be executed using Python. At times, it might be convenient to also compile a model directly in Python. We explores the Python methods to do so. And we also design an interface to package the compilation and operation of the model.

### CODE REVIEW

**Comment by AlexandreEichenberger** [comment]
@Jiaqing-ASU great, will review once #1683 has landed... thanks for working on this

**Comment by Jiaqing-ASU** [comment]
> @Jiaqing-ASU great, will review once #1683 has landed... thanks for working on this

Please allow some time for me to modify some of the code in this PR after #1683 has landed. I will make this PR `Ready to review` once I am done. Many thanks!

**Comment by AlexandreEichenberger** [comment]
PR landed now, so you can update your changes with the latest main branch.

**Comment by AlexandreEichenberger** [comment]
updating branch

File: docs/UsingPyRuntime.md [hunk 1]
```diff
@@ -124,3 +132,137 @@ def set_entry_point(self, name: str):
...
+
+## Running the PyOnnxMlirCompiler interface
+
+An ONNX model can be compiled directly using the `onnx-mlir -O3 --EmitLib` command.
+The resulting library can then be executed using Python as shown in the previous
+sections. At times, it might be convenient to also compile a model directly in Python.
+This section explores the Python methods to do so.
+
+```python
+from PyOnnxMlirCompiler import OnnxMlirCompiler, OnnxMlirTarget, OnnxMlirOption
```
Comment (AlexandreEichenberger) [inline docs/UsingPyRuntime.md:146]: I though we don't need the options anymore. Given that we nearly need to extra interface, I wonder if we could make a subclass of 
```
from PyRuntime import ExecutionSession
```
for example 
```
from PyCompileAndRuntime import CompileExecutionSession
```

where we have a new constructor that takes the .onnx input file, the options, and load and compile.

What do you think @Jiaqing-ASU?

Comment (Jiaqing-ASU) [reply docs/UsingPyRuntime.md:146]: I agree. This is much smarter. Will do it. Thanks.

Comment (AlexandreEichenberger) [reply docs/UsingPyRuntime.md:146]: Thanks for your ongoing work on this, much appreciated

**Comment by AlexandreEichenberger** [comment]
updated the branch to run all cis

**Comment by Jiaqing-ASU** [comment]
> updated the branch to run all cis

@AlexandreEichenberger Thanks for the merge. I guess this will pass the tests since it works in my environment. However, this are still several things of the new constructor need to be modify. I am working on this. Maybe you could review this later. Many thanks.

**Comment by AlexandreEichenberger** [comment]
yes, ping me when you feel its in a state where reviewing it make sense.

File: src/Runtime/CMakeLists.txt [hunk 2]
```diff
@@ -114,3 +114,24 @@ llvm_update_compile_flags(PyRuntime)
...
+  $<TARGET_PROPERTY:onnx,COMPILE_DEFINITIONS>
+  )
+target_include_directories(PyCompileAndRuntime
+  PRIVATE
+  ${ONNX_MLIR_SRC_ROOT}/include
+  )
+target_link_libraries(PyCompileAndRuntime
+  PRIVATE
+  OnnxMlirCompiler
+  ExecutionSession
```
Comment (Jiaqing-ASU) [inline src/Runtime/CMakeLists.txt:131]: @AlexandreEichenberger Hello, Alexandre. I seem to be stuck here. The whole project can compile with no issues. However, it will throw an ImportError with the message `undefined symbol` when trying to load it in Python. The whole error message is:
`ImportError: /workdir/onnx-mlir/docs/mnist_example/PyCompileAndRuntime.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN9onnx_mlir18PyExecutionSession15pySetEntryPointENSt7__cxx1112basic_stringIcSt11char_traitsIcESaIcEEE`

The `undefined symbol` when importing sounds like the built module isn't being linked to the library, which is related to the `target_link_libraries` in the `CMakeLists.txt`. I have tried to remove the `OnnxMlirCompiler` in this `CMakeLists.txt` and it gives the error:
`ImportError: /workdir/onnx-mlir/docs/mnist_example/PyCompileAndRuntime.cpython-38-x86_64-linux-gnu.so: undefined symbol: _ZN9onnx_mlir18PyExecutionSession16pyInputSignatureB5cxx11Ev` which consistent with my guess.

Now the problem is that the error currently I meet is related to the function `pySetEntryPoint()` which is a member in `PyExecutionSession` class. From my understanding, the type `MODULE_LIBRARY` cannot be linked into another target. One may link only to `INTERFACE`, `OBJECT`, `STATIC` or `SHARED` libraries, or to executables with the `ENABLE_EXPORTS` property set. I am not sure whether there is any way to link the `PyRuntime` (type `MODULE_LIBRARY`) which contains the `PyExecutionSession` class with the new class `PyCompileAndRuntime` or maybe I need to find another way to solve this problem. Do you have any ideas or suggestions?

**Comment by AlexandreEichenberger** [comment]
updated to restart also the jenkins

**Comment by Jiaqing-ASU** [comment]
The "undefined symbol" issues have been fixed by making the new class a subclass of onnx_mlir::ExecutionSession instead of onnx_mlir::PyExecutionSession. It now works very well. However, for now, there are about 150 lines of redundant code in onnx_mlir::PyExecutionSession and onnx_mlir::PyCompileExecutionSession. I think this can be fixed later or maybe there are other better solutions to solve this problem. @AlexandreEichenberger Please take a look at all the code and let me know if there is anything that needs to be fixed or you have any suggestions. Many thanks!

**Review by AlexandreEichenberger — COMMENTED** [review]
Gave you comments on the interface itself. Will review code a bit later

**Comment by Jiaqing-ASU** [comment]
> Gave you comments on the interface itself. Will review code a bit later

No problem. Thanks for the helpful comments! I will fix all those comments one by one tomorrow. BTW, do you know why one of the checks `MLIR-Windows-CI` failing? How should I fix that?

File: docs/UsingPyRuntime.md [hunk 3]
```diff
@@ -124,3 +137,156 @@ def set_entry_point(self, name: str):
...
+        Int containing the results. 0 represents successful compilation; others on failure.
+    """
+def get_compiled_file_name(self):
+    """
+    Method to provide the full (absolute or relative) output file name, including
+    its suffix.
+    Returns:
+        String containing the fle name after successful compilation; empty string on failure.
+    """
+def get_error_message(self):
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply docs/UsingPyRuntime.md:286]: Fixed. Use `get_compiled_file_name` for both.

File: docs/UsingPyRuntime.md [hunk 4]
```diff
@@ -1,31 +1,44 @@
...
-Using pybind, a C/C++ binary can be directly imported by the Python interpreter. For onnx-mlir,
-such binary is generated by `PyExecutionSession` (src/Runtime/PyExecutionSession.hpp) and built
-as a shared library to `build/Debug/lib/PyRuntime.cpython-<target>.so`.
+Using pybind, a C/C++ binary can be directly imported by the Python interpreter.
+For onnx-mlir, there are three such libraries, one to compile onnx-mlir models, 
+one to run the models and the other one is to compile and run the models.
 
-## Configuring and using PyRuntime
+1. The library to compile onnx-mlir models is generated
+by `PyOnnxMlirCompiler` (src/Compiler/PyOnnxMlirCompiler.hpp) and build as a shared
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply docs/UsingPyRuntime.md]: Fixed. For now, the ways to use those 3 interfaces in Python are:
`from PyCompile import PyOMCompileSession`
`from PyRuntime import ExecutionSession`
`from PyCompileAndRuntime import PyOMCompileExecutionSession`

File: docs/UsingPyRuntime.md [hunk 5]
```diff
@@ -124,3 +137,156 @@ def set_entry_point(self, name: str):
...
 ```
+
+# Python interface to compile models: PyOnnxMlirCompiler
+
+## Running the PyOnnxMlirCompiler interface
+
+An ONNX model can be compiled directly using the `onnx-mlir -O3 --EmitLib` command.
+The resulting library can then be executed using Python as shown in the previous
+sections. At times, it might be convenient to also compile a model directly in Python.
+This section explores the Python methods to do so.
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply docs/UsingPyRuntime.md]: Fixed.

**Comment by AlexandreEichenberger** [comment]
> No problem. Thanks for the helpful comments! I will fix all those comments one by one tomorrow. BTW, do you know why one of the checks MLIR-Windows-CI failing? How should I fix that?

I don't have access to a window machine; I usually try to see from the failure log, if one is lucky, it is a compile error that one can typically google to find an answer. If one needs to run it, I ask some of the Microsoft folks to help.

**Comment by Jiaqing-ASU** [comment]
> I don't have access to a window machine; I usually try to see from the failure log, if one is lucky, it is a compile error that one can typically google to find an answer. If one needs to run it, I ask some of the Microsoft folks to help. They can often tell you "oh, this is a well known difference, try this" type of comments

Hello @AlexandreEichenberger I took a look at the errors and also searched on Google. The error is from a warning (it shows `the following warning is treated as an error`) which is called [Compiler Warning (level 1) C4530](https://learn.microsoft.com/en-us/cpp/error-messages/compiler-warnings/compiler-warning-level-1-c4530?view=msvc-170). This warning is from the beginning of `PYBIND11_MODULE(PyCompile, m)`. I have compared this part with the PyRuntime and they use exactly the same format. I am not sure whether I need to add `/EHsc` to compile to resolve the warning or not because I did not see this stuff in the previous Runtime CMakeLists. Another interesting thing is that this kind of warning is typically from some code using C++ exception handling but I do not think PyOMCompileSession.hpp has this kind of exception handling. I guess I need some hints or help on this error. Many thanks!

**Comment by AlexandreEichenberger** [comment]
@Jiaqing-ASU if you still have the error, ask @sstamenova to help you, there are ways to disable the warning as errors for sections of the code.

**Comment by Jiaqing-ASU** [comment]
Hello @sstamenova. Could you please help me take a look at the errors in this PR under the MLIR-Windows-CI? This error only shows in MLIR-Windows-CI and I have no idea how to fix the Compiler Warning (level 1) C4530. Your time and help is highly appreciated!

**Comment by sstamenova** [comment]
@Jiaqing-ASU This looks like an issue with the library you are adding. It is using c++ exceptions but it is not being compiled with `/EHsc`. I recommend looking at how mlir itself uses `pybind11_add_module` because it sets a number of compilation flags that are needed for clang, gcc and msvc.

```
C:\PROGRA~2\MICROS~2\2019\ENTERP~1\VC\Tools\MSVC\1429~1.301\bin\Hostx64\x64\cl.exe  /nologo /TP -DONNX_ML=1 -DONNX_MLIR_BUILT_AS_STATIC -DONNX_MLIR_ENABLE_MHLO -DONNX_NAMESPACE=onnx -DPyCompile_EXPORTS -DUNICODE -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -ID:\a\1\llvm-project\build\include -ID:\a\1\onnx-mlir\include -ID:\a\1\onnx-mlir -ID:\a\1\onnx-mlir\build -external:IC:\hostedtoolcache\windows\Python\3.9.13\x64\include -external:ID:\a\1\onnx-mlir\third_party\pybind11\include -external:W0 /DWIN32 /D_WINDOWS   /WX /Zc:inline /Zc:__cplusplus /Oi /bigobj /permissive- /W4 -wd4141 -wd4146 -wd4244 -wd4267 -wd4291 -wd4351 -wd4456 -wd4457 -wd4458 -wd4459 -wd4503 -wd4624 -wd4722 -wd4100 -wd4127 -wd4512 -wd4505 -wd4610 -wd4510 -wd4702 -wd4245 -wd4706 -wd4310 -wd4701 -wd4703 -wd4389 -wd4611 -wd4805 -wd4204 -wd4577 -wd4091 -wd4592 -wd4319 -wd4709 -wd4324 -w14062 -we4238 /Gw /Zc:preprocessor -DSUPPRESS_THIRD_PARTY_WARNINGS /MD /O2 /Ob2   -D_CRT_SECURE_NO_DEPRECATE -D_CRT_SECURE_NO_WARNINGS -D_CRT_NONSTDC_NO_DEPRECATE -D_CRT_NONSTDC_NO_WARNINGS -D_SCL_SECURE_NO_DEPRECATE -D_SCL_SECURE_NO_WARNINGS -DUNICODE -D_UNICODE -D__STDC_CONSTANT_MACROS -D__STDC_FORMAT_MACROS -D__STDC_LIMIT_MACROS -UNDEBUG /bigobj /MP /GL -std:c++17 /showIncludes /Fosrc\Compiler\CMakeFiles\PyCompile.dir\PyOMCompileSession.cpp.obj /Fdsrc\Compiler\CMakeFiles\PyCompile.dir\ /FS -c D:\a\1\onnx-mlir\src\Compiler\PyOMCompileSession.cpp
D:\a\1\onnx-mlir\src\Compiler\PyOMCompileSession.hpp(48): error C2220: the following warning is treated as an error
D:\a\1\onnx-mlir\src\Compiler\PyOMCompileSession.hpp(48): warning C4530: C++ exception handler used, but unwind semantics are not enabled. Specify /EHsc
```

**Comment by Jiaqing-ASU** [comment]
> @Jiaqing-ASU This looks like an issue with the library you are adding. It is using c++ exceptions but it is not being compiled with `/EHsc`. I recommend looking at how mlir itself uses `pybind11_add_module` because it sets a number of compilation flags that are needed for clang, gcc and msvc.

Thanks @sstamenova for your suggestion. I just would like to make sure I understand your suggestion correctly. From my understanding, `how mlir itself uses pybind11_add_module` refers the [PyRuntime](https://github.com/onnx/onnx-mlir/blob/main/src/Runtime/PyExecutionSession.cpp) code and I did compare my code with the PyRuntime as well as the `CMakeLists.txt` under `Runtime` folder and `Compiler` folder line by line. The only difference I have noticed is that in PyRuntime, there is a piece of code intercepts some warnings:
```
#include "src/Support/SuppressWarnings.h"
SUPPRESS_WARNINGS_PUSH
#include "onnx/onnx_pb.h"
SUPPRESS_WARNINGS_POP
```
I have added this part to my code as well and there is no errors in my Linux docker-dev environment. Hope this can work on the MLIR-Windows-CI. @AlexandreEichenberger Could you please help me check whether the code right now can work. Thanks!

If this still does not work, the warning might not be from here. Inspired by @AlexandreEichenberger, he mentioned that `there are ways to disable the warning as errors for sections of the code.` I am thinking about adding the following code in the `PyOMCompileSession.hpp` if the previous solution does not work:
```
#pragma warning(push)
#pragma warning(disable : C4530)
#pragma warning(pop)
```
Even if this method can solve the issue, this is not the method I like/prefer to use. I want to know the root of this problem. Please let me know if you have any new insights or suggestions? Thanks very much!

**Comment by sstamenova** [comment]
@Jiaqing-ASU : No, this was not my suggestion at all. We do *not* want to suppress this warning, we want to fix it instead. We also don't want to add an onnx dependency if one is not needed, so I think you need to revert your latest change.

My recommendation was to look at mlir itself: https://github.com/llvm/llvm-project/tree/main/mlir and see how they use the pybind modules. Their usage of pybind modules is a lot more sophisticated and addresses the issue you are running into (in addition to a couple of others):

https://github.com/llvm/llvm-project/blob/24989646a5216cc6f8c6d336a17e7cb2f5c60454/mlir/cmake/modules/AddMLIRPython.cmake#L610

**Comment by Jiaqing-ASU** [comment]
@sstamenova Thank you for your quick reply. I think I understand your advice now. It means actually even the `PyRuntime`, which has been used for a long time before, did not use the correct method to handle this kind of warning. I think I will take a closer look at how `mlir` use the pybind modules and fix those stuff all properly. Thank you very much for your help!

File: src/Runtime/PyOMCompileExecutionSession.hpp [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+  std::string pyGetCompiledFileName();
+  std::string pyGetErrorMessage();
+  int64_t pyGetCompiledResult();
+  std::vector<std::string> pyQueryEntryPoints();
+  void pySetEntryPoint(std::string entryPointName);
+  std::vector<py::array> pyRun(const std::vector<py::array> &inputsPyArray);
+  std::string pyInputSignature();
+  std::string pyOutputSignature();
+
+private:
```
Comment (AlexandreEichenberger) [inline src/Runtime/PyOMCompileExecutionSession.hpp:41]: Would we not have less code replication if this combined Compile and execution did not inherit from either, but has inside one object `PyOMCompileSession` and one `ExecutionSession` and then we just forward the functions to the right object? Ideally there should be no new code here, right?

Comment (Jiaqing-ASU) [reply src/Runtime/PyOMCompileExecutionSession.hpp:41]: > Would we not have less code replication if this combined Compile and execution did not inherit from either, but has inside one object `PyOMCompileSession` and one `ExecutionSession` and then we just forward the functions to the right object? Ideally there should be no new code here, right?

Yes. There should be no new code.

Comment (AlexandreEichenberger) [reply src/Runtime/PyOMCompileExecutionSession.hpp:41]: Cool, hopefully a quick refactor and we can declare success :-)

File: docs/mnist_example/README.md [hunk 7]
```diff
@@ -11,7 +11,7 @@ Table of Contents
       * [Write a Python Driver Code](#write-a-Python-driver-code)
       * [Write a Java Driver Code](#write-a-java-driver-code)
 
-# Train Model in PyTorch, Compile using ONNX-MLIR
+#Train Model in PyTorch, Compile using ONNX - MLIR
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply docs/mnist_example/README.md:14]: I guess this might be because I have applied the `clang format` to this README file. Fixed.

File: src/Compiler/PyOMCompileSession.cpp [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+SUPPRESS_WARNINGS_POP
+
+#include "PyOMCompileSession.hpp"
+
+namespace onnx_mlir {
+
+int64_t PyOMCompileSession::pyCompileFromFile(std::string flags) {
+  const char *outputName, *errorMsg;
+  int64_t rc;
+  rc = omCompileFromFile(
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply src/Compiler/PyOMCompileSession.cpp:36]: Good suggestion. Fixed.

File: src/Compiler/PyOMCompileSession.cpp [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+    outputFileName = std::string();
+  }
+  return rc;
+}
+
+int64_t PyOMCompileSession::pyCompileFromArray(
+    std::string outputBaseName, EmissionTargetType emissionTarget) {
+  const char *outputName, *errorMsg;
+  int64_t rc;
+  rc = omCompileFromArray(inputBuffer, inputBufferSize, outputBaseName.c_str(),
```
[the remark this answers is not in the document]
Comment (Jiaqing-ASU) [reply src/Compiler/PyOMCompileSession.cpp:56]: Good suggestion. Fixed.

File: src/Runtime/PyOMCompileExecutionSession.hpp (hunk 6 above)
Comment (Jiaqing-ASU) [reply src/Runtime/PyOMCompileExecutionSession.hpp:41]: Done.

**Comment by Jiaqing-ASU** [comment]
Hello @sstamenova. Thanks for your guidance to solve `/EHsc` issue! It works well in my previous version of code. But, right now, I meet a new issue. MLIR-Windows-CI gives new errors (`error LNK2001: unresolved external symbol`) which seems like I need to link the header files where these functions are located in the `target_link_libraries`.

```
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl onnx_mlir::PyExecutionSession::pyOutputSignature(void)" (?pyOutputSignature@PyExecutionSession@onnx_mlir@@QEAA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl onnx_mlir::PyExecutionSession::pyInputSignature(void)" (?pyInputSignature@PyExecutionSession@onnx_mlir@@QEAA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::vector<class pybind11::array,class std::allocator<class pybind11::array> > __cdecl onnx_mlir::PyExecutionSession::pyRun(class std::vector<class pybind11::array,class std::allocator<class pybind11::array> > const &)" (?pyRun@PyExecutionSession@onnx_mlir@@QEAA?AV?$vector@Varray@pybind11@@V?$allocator@Varray@pybind11@@@std@@@std@@AEBV34@@Z)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: void __cdecl onnx_mlir::PyExecutionSession::pySetEntryPoint(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (?pySetEntryPoint@PyExecutionSession@onnx_mlir@@QEAAXV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::vector<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,class std::allocator<class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > > > __cdecl onnx_mlir::PyExecutionSession::pyQueryEntryPoints(void)" (?pyQueryEntryPoints@PyExecutionSession@onnx_mlir@@QEAA?AV?$vector@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@V?$allocator@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@2@@std@@XZ)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: __cdecl onnx_mlir::PyExecutionSession::PyExecutionSession(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,bool)" (??0PyExecutionSession@onnx_mlir@@QEAA@V?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@_N@Z)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl onnx_mlir::PyOMCompileSession::pyGetErrorMessage(void)" (?pyGetErrorMessage@PyOMCompileSession@onnx_mlir@@QEAA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> > __cdecl onnx_mlir::PyOMCompileSession::pyGetCompiledFileName(void)" (?pyGetCompiledFileName@PyOMCompileSession@onnx_mlir@@QEAA?AV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@XZ)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: __int64 __cdecl onnx_mlir::PyOMCompileSession::pyCompileFromArray(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >,enum onnx_mlir::EmissionTargetType)" (?pyCompileFromArray@PyOMCompileSession@onnx_mlir@@QEAA_JV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@W4EmissionTargetType@2@@Z)
PyOMCompileExecutionSession.cpp.obj : error LNK2001: unresolved external symbol "public: __int64 __cdecl onnx_mlir::PyOMCompileSession::pyCompileFromFile(class std::basic_string<char,struct std::char_traits<char>,class std::allocator<char> >)" (?pyCompileFromFile@PyOMCompileSession@onnx_mlir@@QEAA_JV?$basic_string@DU?$char_traits@D@std@@V?$allocator@D@2@@std@@@Z)
Release\lib\PyCompileAndRuntime.cp39-win_amd64.pyd : fatal error LNK1120: 10 unresolved externals
```

As those files cannot be linked into `target_link_libraries`, I would like to ask whether this is any other solutions to solve this error? If not, I will just revert to my previous code. Beacuse the reason for me to refactor the code just to reduce some duplicate code and my previous code works well with no warnings or errors. Thanks for the help!

File: src/Runtime/CMakeLists.txt [hunk 10]
```diff
@@ -97,12 +97,18 @@ set_target_properties(ExecutionSession
...
+  $<$<OR:$<CXX_COMPILER_ID:Clang>,$<CXX_COMPILER_ID:AppleClang>,$<CXX_COMPILER_ID:GNU>>:-frtti -fexceptions>
+  $<$<CXX_COMPILER_ID:MSVC>:/EHsc /GR>
+  )
 target_compile_definitions(PyRuntime
   PRIVATE
   $<TARGET_PROPERTY:onnx,COMPILE_DEFINITIONS>
   )
 target_include_directories(PyRuntime
   PRIVATE
+  ${ONNX_MLIR_SRC_ROOT}/include
```
Comment (sstamenova) [inline src/Runtime/CMakeLists.txt:111]: Why is this needed now?

File: src/Compiler/CMakeLists.txt [hunk 11]
```diff
@@ -182,3 +185,28 @@ add_onnx_mlir_library(OnnxMlirCompiler
...
+  $<$<CXX_COMPILER_ID:MSVC>:/EHsc /GR>
+  )
+target_compile_definitions(PyCompile
+  PRIVATE
+  $<TARGET_PROPERTY:onnx,COMPILE_DEFINITIONS>
+  )
+target_include_directories(PyCompile
+  PRIVATE
+  $<TARGET_PROPERTY:onnx,INCLUDE_DIRECTORIES>
+  ${ONNX_MLIR_SRC_ROOT}/include
```
Comment (sstamenova) [inline src/Compiler/CMakeLists.txt:203]: This should not be needed

File: src/Runtime/CMakeLists.txt [hunk 12]
```diff
@@ -114,3 +120,30 @@ llvm_update_compile_flags(PyRuntime)
...
+target_compile_definitions(PyCompileAndRuntime
+  PRIVATE
+  $<TARGET_PROPERTY:onnx,COMPILE_DEFINITIONS>
+  )
+target_include_directories(PyCompileAndRuntime
+  PRIVATE
+  ${ONNX_MLIR_SRC_ROOT}/include
+  $<TARGET_PROPERTY:onnx,INCLUDE_DIRECTORIES>
+  )
+target_link_libraries(PyCompileAndRuntime
```
Comment (sstamenova) [inline src/Runtime/CMakeLists.txt:140]: Why aren't you linking this against PyCompile  and PyRuntime?

Comment (Jiaqing-ASU) [reply src/Runtime/CMakeLists.txt:140]: Because `PyCompile` and `PyRuntime` are not normal C++ libraries. Furthermore, they are not even normal `shared libraries` which means this [solution](https://github.com/pybind/cmake_example/issues/11) I have tried before does not work. I have also tried another [solution](https://github.com/pybind/pybind11/issues/2056) to combine two pybind11-wrapped projects. It also does not work either as it is a little bit different from our situation. We need to have two objects of PyCompile and PyRuntime inside the new class and then use the functions inside the objects instead of just calling functions from another pybind11-generated Python binding. Also, I did not get enough information from the [llvm-project pybind11_add_module](https://github.com/llvm/llvm-project/blob/24989646a5216cc6f8c6d336a17e7cb2f5c60454/mlir/cmake/modules/AddMLIRPython.cmake#L638). Maybe I am looking into some wrong places.
